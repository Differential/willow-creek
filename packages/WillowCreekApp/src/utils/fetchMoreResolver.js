import get from 'lodash/get';
import set from 'lodash/fp/set';

export default function fetchMoreResolver({
  collectionName,
  variables,
  fetchMore,
  data,
} = {}) {
  return () => {
    const pageInfoPath = `${collectionName}.pageInfo`;
    const edgePath = `${collectionName}.edges`;

    const after = get(data, `${pageInfoPath}.endCursor`);
    if (!after) return;

    fetchMore({
      variables: { ...variables, after },
      updateQuery: (previousResult, { fetchMoreResult }) => {
        // combine previous data and fetchMore data
        const newDataWithPageInfo = set(
          pageInfoPath,
          {
            ...get(previousResult, pageInfoPath, {}),
            ...get(fetchMoreResult, pageInfoPath, {}),
          },
          previousResult
        );
        const newDataWithAdditionalItems = set(
          edgePath,
          [
            ...get(previousResult, edgePath, []),
            ...get(fetchMoreResult, edgePath, []),
          ],
          newDataWithPageInfo
        );

        return newDataWithAdditionalItems;
      },
    });
  };
}
