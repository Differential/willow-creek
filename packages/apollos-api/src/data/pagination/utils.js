// eslint-disable-next-line import/prefer-default-export
export const withEdgePagination = async ({ edges }) => {
  const result = await edges;
  const startCursor = result.length ? result[0].cursor : null;
  const endCursor = result.length ? result[result.length - 1].cursor : null;
  return { startCursor, endCursor };
};
