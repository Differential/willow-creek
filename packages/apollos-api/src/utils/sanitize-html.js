import sanitizeHtml from 'sanitize-html';

const allowedTags = [
  'h1',
  'h2',
  'h3',
  'h4',
  'h5',
  'h6',
  'blockquote',
  'p',
  'a',
  'ul',
  'ol',
  'li',
  'b',
  'i',
  'strong',
  'em',
  'br',
  'caption',
  'img',
];

const allowedAttributes = {
  a: ['href'],
  img: ['src'],
};

// A very picky HTML sanitizer
export default function(dirty) {
  return sanitizeHtml(dirty, {
    allowedTags,
    allowedAttributes,
  });
}
