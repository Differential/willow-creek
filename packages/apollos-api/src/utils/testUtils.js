import { KeyValueCache } from 'apollo-server-caching';
import {
  context as getContext,
  dataSources as getDataSources,
} from '../data/index';

export function getTestContext(req) {
  const context = getContext(req);
  const dataSources = getDataSources();
  // Apollo Server does this internally.
  Object.values(dataSources).forEach((dataSource) => {
    if (dataSource.initialize) {
      dataSource.initialize({ context, cache: KeyValueCache });
    }
  });
  context.dataSources = dataSources;
  return context;
}

export const buildGetMock = (response, dataSource) => {
  const get = jest.fn();
  if (Array.isArray(response) && Array.isArray(response[0])) {
    response.forEach((responseVal) => {
      get.mockReturnValueOnce(
        new Promise((resolve) => resolve(dataSource.normalize(responseVal)))
      );
    });
  }
  get.mockReturnValue(
    new Promise((resolve) => resolve(dataSource.normalize(response)))
  );
  return get;
};
