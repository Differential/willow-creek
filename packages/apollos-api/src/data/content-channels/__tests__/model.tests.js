import ApollosConfig from '@apollosproject/config';
import { buildGetMock } from '../../../utils/testUtils';
import ContentChannelDataSource from '../data-source';

ApollosConfig.loadJs({
  ROCK: {
    API_URL: 'https://apollosrock.newspring.cc/api',
    API_TOKEN: 'some-rock-token',
    IMAGE_URL: 'https://apollosrock.newspring.cc/GetImage.ashx',
  },
});

describe('ContentChannelModel', () => {
  it('constructs', () => {
    expect(new ContentChannelDataSource()).toBeTruthy();
  });
  it('gets all', () => {
    const dataSource = new ContentChannelDataSource();
    dataSource.get = buildGetMock([{ Id: 1 }, { Id: 2 }], dataSource);
    const result = dataSource.all();
    expect(result).resolves.toMatchSnapshot();
    expect(dataSource.get.mock.calls).toMatchSnapshot();
  });

  it('gets by id', () => {
    const dataSource = new ContentChannelDataSource();
    dataSource.get = buildGetMock([{ Id: 1 }], dataSource);
    const result = dataSource.getFromId(1);
    expect(result).resolves.toMatchSnapshot();
    expect(dataSource.get.mock.calls).toMatchSnapshot();
  });
});
