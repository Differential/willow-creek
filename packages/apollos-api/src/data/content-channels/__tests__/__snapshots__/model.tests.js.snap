// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentChannelModel gets all 1`] = `
Array [
  Array [
    "ContentChannels?%24expand=ChildContentChannels",
    Object {},
    Object {},
  ],
]
`;

exports[`ContentChannelModel gets all 2`] = `
Array [
  Object {
    "id": 1,
  },
  Object {
    "id": 2,
  },
]
`;

exports[`ContentChannelModel gets by id 1`] = `
Array [
  Array [
    "ContentChannels?%24filter=Id%20eq%201&%24expand=ChildContentChannels&%24top=1",
    Object {},
    Object {},
  ],
]
`;

exports[`ContentChannelModel gets by id 2`] = `
Object {
  "id": 1,
}
`;
